cloud/amazon/s3_bucket.py

```diff
@@ -146,73 +143,57 @@ def create_bucket(connection, module, location):
-    if not versioning_status and versioning:
-        try:
-            bucket.configure_versioning(versioning)
-            changed = True
-            versioning_status = bucket.get_versioning_status()
-        except S3ResponseError, e:
-            module.fail_json(msg=e.message)
-    elif not versioning_status and not versioning:
-        # do nothing
-        pass
-    else:
-        if versioning_status['Versioning'] == "Enabled" and not versioning:
-            bucket.configure_versioning(versioning)
-            changed = True
-            versioning_status = bucket.get_versioning_status()
-        elif ( (versioning_status['Versioning'] == "Disabled" and versioning) or (versioning_status['Versioning'] == "Suspended" and versioning) ):
-            bucket.configure_versioning(versioning)
-            changed = True
-            versioning_status = bucket.get_versioning_status()
-    
+    if versioning_status or versioning:
+        if not versioning_status or \
+           (not versioning and versioning_status['Versioning'] == "Enabled") or \
+           (versioning and versioning_status['Versioning'] in ["Disabled", "Suspended"]):
+            try:
+                bucket.configure_versioning(versioning)
+                changed = True
+                versioning_status = bucket.get_versioning_status()
+            except S3ResponseError, e:
+                    module.fail_json(msg=e.message)
+
     # Requester pays
     requester_pays_status = get_request_payment_status(bucket)
     if requester_pays_status != requester_pays:
         if requester_pays:
-            bucket.set_request_payment(payer='Requester')
-            changed = True
-            requester_pays_status = get_request_payment_status(bucket)
+            payer='Requester'
         else:
-            bucket.set_request_payment(payer='BucketOwner')
-            changed = True
-            requester_pays_status = get_request_payment_status(bucket)
+            payer='BucketOwner'
+        bucket.set_request_payment(payer=payer)
+        changed = True
+        requester_pays_status = get_request_payment_status(bucket)
 
-    # Policy        
+    # Policy
     try:
         current_policy = bucket.get_policy()
     except S3ResponseError, e:
         if e.error_code == "NoSuchBucketPolicy":
             current_policy = None
         else:
             module.fail_json(msg=e.message)
-    
-    if current_policy is not None and policy is not None:
-        if policy is not None:
-            policy = json.dumps(policy)
-            
-        if json.loads(current_policy) != json.loads(policy):
+
+    if policy is not None:
+        # Deal with policy if either JSON formatted string or just data structure
+        if isinstance(policy, basestring):
+            compare_policy = json.dumps(policy)
+            load_policy = policy
+        else:
+            compare_policy = policy
+            load_policy = json.loads(policy)
```

Comment:
This is reversed (Note I also threw in strip() to take care of the issue reported on the mailing list):
``` python
if isinstance(policy, basestring):
    compare_policy = json.loads(policy)
    load_policy = policy.strip()
else:
    compare_policy = policy
    load_policy = json.dumps(policy)
```